themes/default/content/learn/abstraction-encapsulation/component-resources/index.md

```diff
@@ -0,0 +1,113 @@
+_Component Resource_ in Pulumi's terminology.
+
+## Deciding to create a component resource
+
+Why would you make a component resource instead of just a "simple" logical
+grouping that you can call and use? A component resource shows up in the Pulumi
+ecosystem just like any other resource. That means it has a trackable state,
+appears in diffs, and has a name field you can reference that refers to the
+entire grouping.
+
+We've actually already started creating a component resource in the last part
+when we made a new class that built up an `s3.Bucket` and an `s3.BucketPolicy`
+with a `define_policy` function. Let's now turn that logical grouping into an
+actual component resource.
+
+## Converting to a component resource
+
+When we're converting to a component resource, we're subclassing (or composing
+from an anonymous field, if you're into Go) the `ComponentResource` so that our
+new component resource can get all of the lovely benefits of a resource (state
+tracking, diffs, name fields, etc.) that other resources have.
+
+In Python, we subclass by using `super()` in the initialization of the class.
+This call ensures that Pulumi registers the component resource as a resource
+properly.
+
+```python
+import ...
+
+
+class OurBucketComponent(pulumi.ComponentResource):
+    def __init__(self, name, policy_name='default', opts=None):
+        super().__init__('pkg:index:OurBucketComponent', name, None, opts)
+        child_opts = pulumi.ResourceOptions(parent=self)
+        policy_list = {
+            'default': '{...}',
+            'locked' : '{...}',
+            'permissive': '{...}'
+        }
+        bucket = aws.s3.bucket(name)
+        def define_policy(policy_name):
+            try:
+                if policy_name in policy_list.keys():
+                    json_data = policy_list[f"{policy_name}"]
+                return json.dumps(json_data)
+            except Exception as err:
+                raise err
+        bucket_policy = aws_classic.s3.BucketPolicy(
+            name,
+            bucket=bucket.id,
+            policy=bucket.arn.apply(lambda arn: define_policy(policy_name)),
+            opts=pulumi.ResourceOptions(parent=bucket)
+        )
+        self.register_outputs({
+            "bucket_name": bucket.bucket_name
+        })
+```
+
+Within `super()`'s init, we pass in a name for the resource, which we recommend
+being of the form `<package>:<module>:<type>` to avoid type conflicts since it's
+being registered alongside other resources like the Bucket resource we're
+calling (`aws:s3:Bucket`).
+
+That last call, `self.register_outputs({})`, passes Pulumi the expected outputs
+so Pulumi can read the results of the creation or update of a component resource
+just like any other resource, so don't forget it! You can register default
+outputs using this call, as well. It's not hard to imagine we will always want
+the bucket name for our use case, so we pass that in as an always-given output
+for our component resource.
```

Comment:
It'd be _really_ nice to see how this looks in an output or in the Console somehow, vs. how things look without it.

Reply:
I think I'm going to get this out with the next round when we've got the TS support in. I'm just not sure I'm doing it right, and there's not a lot of Python examples :/